lib/extras.py

```diff
@@ -781,19 +782,26 @@ def wait_select(conn):
             continue
 
 
+@contextlib.contextmanager
 def _solve_conn_curs(conn_or_curs):
     """Return the connection and a DBAPI cursor from a connection or cursor."""
     if conn_or_curs is None:
         raise psycopg2.ProgrammingError("no connection or cursor provided")
 
     if hasattr(conn_or_curs, 'execute'):
         conn = conn_or_curs.connection
-        curs = conn.cursor(cursor_factory=_cursor)
+        curs = conn_or_curs
```

Comment:
So it looks like this is intentionally always creating a new cursor and not using the passed in the cursor. Creating a new cursor is required for `test_non_dbapi_connection` to pass.

As a new cursor is always created, I think it should always be closed as well. Do you agree?

Reply:
Yes, true. It works that way because sometimes user pass non-standard cursors, and if a non-DBAPI-compliant one is received the user function may fail (see #114, 21d323d2). The output of _solve_conn_curs() is now always a base class cursor.

Reply:
And yes, of course a cursor we own can be safely closed (it won't be a server-side cursor, closing it won't cause any side effect).